File: Core/browser/src/main/java/com/klavor/browser/JsBridge.java
package com.klavor.browser;

import android.util.Log;
import android.webkit.WebView;

import com.klavor.browser.jsbridge.JsExcutor;
import com.klavor.browser.jsbridge.api.JsApi;
import com.klavor.browser.jsbridge.api.JsApiMapping;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Set;

public class JsBridge {
    private static String javascript;

    public static void inject(WebView view) {
        if (javascript == null) {
            synchronized (JsBridge.class) {
                if (javascript == null) {
                    javascript = buildJavascript();
                }
            }
        }
        JsExcutor.excuteJs(view, javascript);
    }

    private static String buildJavascript() {
        StringBuffer sb = new StringBuffer();
        sb.append("window.jsBridge = {\n" +
                "    run : function(host, path, arg) {\n" +
                "        location.href = \"jsBridge://\" + host + \"/\" + path + \"?params=\" + JSON.stringify(arg)\n" +
                "    }\n" +
                "}\n");

        HashMap<String, Class<? extends JsApi>> apiMapping = JsApiMapping.getApiMapping();
        Set<String> keySet = apiMapping.keySet();
        sb.append("window.jsApi = {");
        for (String key : keySet) {
            sb.append(key).append(":{");
            Class<? extends JsApi> clazz = apiMapping.get(key);
            Method[] declaredMethods = clazz.getDeclaredMethods();
            if (declaredMethods != null) {
                for (Method method : declaredMethods) {
                    boolean annotationPresent = method.isAnnotationPresent(JsMehod.class);
                    if (annotationPresent) {
                        sb.append(method.getName()).append(":function(arg){");
                        sb.append("jsBridge.run('" + key + "', '" + method.getName() + "', arg)");
                        sb.append("},");
                    }
                }
                sb.deleteCharAt(sb.length() - 1);
            }
            sb.append("},");
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append("}");
        Log.d("cmf", sb.toString());
        return sb.toString();
    }
}

File: Core/browser/src/main/assets/js/jsBridge.js
window.jsBridge = {
    run : function(host, path, arg, id) {
        location.href = "jsBridge://" + host + "/" + path + "?params=" + JSON.stringify(arg) + "&id=" + id;
    },

    doJs : function(host, path, arg) {
        var id = window.callback.add(function(rtn, err) {
            if (rtn.status == 0) {
                arg.cancel(rtn);
            } else if (rtn.status == -1) {
                arg.error(rtn, err);
            } else if (rtn.status == 1) {
                arg.success(rtn);
            }
        })
        this.run(host, path, arg, id);
    }
}

window.callback = {
    list : {},
    id : 1,
    add : function(arg) {
        this.list[this.id] = arg;
        return this.id++;
    },
    cb : function(id, rtn, err, del) {
        this.list[id] && this.list[id](rtn, err);
        if (del) {
            delete this.list[id];
        }
    }
}
